src/textual/widgets/_tree.py

```diff
@@ -330,34 +336,101 @@ def add(
         Args:
             label: The new node's label.
             data: Data associated with the new node.
+            before: Optional index or `TreeNode` to add the node before.
+            after: Optional index or `TreeNode` to add the node after.
             expand: Node should be expanded.
             allow_expand: Allow use to expand the node via keyboard or mouse.
 
         Returns:
             A new Tree node
+
+        Raises:
+            AddNodeError: If there is a problem with the addition request.
+
+        Note:
+            Only one of `before` or `after` can be provided. If both are
+            provided a `AddNodeError` will be raised.
         """
+        if before is not None and after is not None:
+            raise AddNodeError("Unable to add a node both before and after a node")
+
+        insert_index: int = len(self.children)
+
+        if before is not None:
+            if isinstance(before, int):
+                insert_index = before
+            elif isinstance(before, TreeNode):
+                try:
+                    insert_index = self.children.index(before)
+                except ValueError:
+                    raise AddNodeError(
+                        "Request to add before an unknown node that is not a child of this node"
```

Comment:
Struggled to mentally parse the error message a bit here.

```suggestion
                        "The node specified for 'before' does not exist in this Tree."
```

Reply:
Actually on reading the code in more detail I think my suggestion isn't clear.

Maybe something like `f"cannot insert before {before}, as this node only has {len(children)} children"`?

Reply:
Haha yeah, I realised that after posting. See my other comment 😄

Reply:
> f"cannot insert before {before}, as this node only has {len(children)} children"

Sorry but I'm not sure that's quite right either. The length of the children isn't relevant, rather that the `TreeNode` passed as the `before` isn't a child node of this node?

Reply:
ah sorry, i thought we were in the `isinstance(before, int)` branch 🤦 

how about:

```suggestion
                        "The node specified for `before` is not a child of this node"
```

Reply:
Sorry also just to clarify, if the `before`/`after` argument is an index that is out of range, this also won't raise an error (the same as Python's `insert`)

Reply:
Ah ok - that seems fine to me (I wasn't aware that's how insert worked)

Reply:
Thanks, that wording is much better! Updated in 04ac1ca (sorry I tried adding you as co-author but always forget that second hyphen!)